File: old-scala-codegen/src/bsp/codegen/ir/IR.scala
package bsp.codegen.ir

import software.amazon.smithy.model.shapes.ShapeId

// This file contains an internal representation specifically tailored for the needs and idioms
// of bsp4j. It is not meant to be reused for Scala code-generation, or any other language, for that matter.

sealed trait Def {
  def members: List[ShapeId] = List.empty

  def shapeId: ShapeId

  def hints: List[Hint]
}

object Def {
  final case class Alias(shapeId: ShapeId, aliasedType: Type, hints: List[Hint]) extends Def

  final case class Structure(
      shapeId: ShapeId,
      fields: List[Field],
      hints: List[Hint],
      associatedDataKinds: List[PolymorphicDataKind]
  ) extends Def {
    override def members: List[ShapeId] =
      fields.flatMap(_.tpe.members()) ++ associatedDataKinds.map(_.shapeId)
  }

  final case class OpenEnum[A](
      shapeId: ShapeId,
      enumType: EnumType[A],
      values: List[EnumValue[A]],
      hints: List[Hint]
  ) extends Def

  final case class ClosedEnum[A](
      shapeId: ShapeId,
      enumType: EnumType[A],
      values: List[EnumValue[A]],
      hints: List[Hint]
  ) extends Def

  final case class Service(shapeId: ShapeId, operations: List[Operation], hints: List[Hint])
      extends Def
}

sealed trait JsonRPCMethodType extends Product with Serializable

object JsonRPCMethodType {
  case object Request extends JsonRPCMethodType

  case object Notification extends JsonRPCMethodType
}

final case class Operation(
    shapeId: ShapeId,
    inputType: Type,
    outputType: Type,
    jsonRPCMethodType: JsonRPCMethodType,
    jsonRPCMethod: String,
    hints: List[Hint]
) {
  def name: String = shapeId.getName()
}

sealed trait EnumType[A]

object EnumType {
  case object IntEnum extends EnumType[Int]

  case object StringEnum extends EnumType[String]
}

final case class EnumValue[A](name: String, value: A, hints: List[Hint])

final case class Field(
    name: String,
    tpe: Type,
    required: Boolean,
    jsonRename: Option[String],
    hints: List[Hint]
)

final case class Alternative(name: String, typ: Type, hints: List[Hint])

sealed trait Primitive

object Primitive {
  case object PUnit extends Primitive

  case object PBool extends Primitive

  case object PString extends Primitive

  case object PInt extends Primitive

  case object PLong extends Primitive

  case object PFloat extends Primitive

  case object PDouble extends Primitive

  case object PDocument extends Primitive

  case object PTimestamp extends Primitive
}

sealed trait Type {
  def isPrimitive(primitive: Primitive): Boolean = false

  def members(): List[ShapeId]
}

object Type {
  case class TCollection(member: Type) extends Type {
    override def members(): List[ShapeId] = member.members()
  }

  case class TSet(member: Type) extends Type {
    override def members(): List[ShapeId] = member.members()
  }

  case class TMap(key: Type, value: Type) extends Type {
    override def members(): List[ShapeId] = key.members() ++ value.members()
  }

  case class TRef(shapeId: ShapeId) extends Type {
    override def members(): List[ShapeId] = List(shapeId)
  }

  case class TPrimitive(prim: Primitive, shapeId: ShapeId) extends Type {
    override def isPrimitive(primitive: Primitive): Boolean = prim == primitive

    override def members(): List[ShapeId] = List(shapeId)
  }

  case class TUntaggedUnion(tpes: List[Type]) extends Type {
    override def members(): List[ShapeId] = tpes.flatMap(_.members())
  }

  val TUnit = TPrimitive(Primitive.PUnit, ShapeId.from("smithy.api#void"))
  val TBool = TPrimitive(Primitive.PBool, ShapeId.from("smithy.api#Boolean"))
  val TString = TPrimitive(Primitive.PString, ShapeId.from("smithy.api#String"))
  val TInt = TPrimitive(Primitive.PInt, ShapeId.from("smithy.api#Integer"))
  val TLong = TPrimitive(Primitive.PLong, ShapeId.from("smithy.api#Long"))
  val TFloat = TPrimitive(Primitive.PFloat, ShapeId.from("smithy.api#Float"))
  val TDouble = TPrimitive(Primitive.PDouble, ShapeId.from("smithy.api#Double"))
  val TDocument = TPrimitive(Primitive.PDocument, ShapeId.from("smithy.api#Document"))
  val TTimestamp = TPrimitive(Primitive.PTimestamp, ShapeId.from("smithy.api#Timestamp"))
}

sealed trait Hint

object Hint {
  case class Documentation(string: String) extends Hint
  case class Deprecated(message: String) extends Hint
}

case class PolymorphicDataKind(kind: String, shapeId: ShapeId)

File: codegen/src/main/kotlin/bsp/codegen/common/ir/IR.kt
package bsp.codegen.common.ir

import software.amazon.smithy.model.shapes.ShapeId

sealed interface Def {
  val name
    get() = shapeId.name

  val shapeId: ShapeId
  val hints: List<Hint>

  data class Alias(
      override val shapeId: ShapeId,
      val aliasedType: Type,
      override val hints: List<Hint>
  ) : Def

  data class Service(
      override val shapeId: ShapeId,
      val operations: List<Operation>,
      override val hints: List<Hint>
  ) : Def

  data class Structure(
      override val shapeId: ShapeId,
      val fields: List<Field>,
      override val hints: List<Hint>,
  ) : Def

  data class OpenEnum<A>(
      override val shapeId: ShapeId,
      val enumType: EnumType<A>,
      val values: List<EnumValue<A>>,
      override val hints: List<Hint>
  ) : Def

  data class ClosedEnum<A>(
      override val shapeId: ShapeId,
      val enumType: EnumType<A>,
      val values: List<EnumValue<A>>,
      override val hints: List<Hint>
  ) : Def

  data class DataKinds(
      override val shapeId: ShapeId,
      val kindsEnumId: ShapeId,
      val kinds: List<PolymorphicDataKind>,
      override val hints: List<Hint>
  ) : Def

  data class UntaggedUnion(
      override val shapeId: ShapeId,
      val members: List<Type>,
      override val hints: List<Hint>
  ) : Def
}

fun Def.referencedShapeIds(): List<ShapeId> =
    when (this) {
      is Def.Alias ->
          when (aliasedType) {
            is Type.Ref -> listOf(aliasedType.shapeId)
            else -> emptyList()
          }
      is Def.Service ->
          operations
              .flatMap { listOf(it.inputType, it.outputType) }
              .flatMap { it.referencedShapeIds() }
      is Def.Structure -> fields.flatMap { it.type.referencedShapeIds() }
      is Def.OpenEnum<*> -> emptyList()
      is Def.ClosedEnum<*> -> emptyList()
      is Def.DataKinds -> kinds.map { it.shape }
      is Def.UntaggedUnion -> members.flatMap { it.referencedShapeIds() }
    }

sealed interface Type {
  // primitive types
  object Unit : Type

  object Bool : Type

  object String : Type

  object Int : Type

  object Long : Type

  object Json : Type

  // collections
  data class Set(val member: Type) : Type

  data class List(val member: Type) : Type

  data class Map(val key: Type, val value: Type) : Type

  // references
  data class Ref(val shapeId: ShapeId) : Type

  // Def as type
  data class UntaggedUnion(val members: kotlin.collections.List<Type>) : Type
}

// fun Type.referencedShapeIds(): List<ShapeId> {
//  val set = mutableSetOf<ShapeId>()
//
//  fun go(type: Type) {
//    when (type) {
//      is Type.Ref -> set.add(type.shapeId)
//      is Type.Set -> go(type.member)
//      is Type.List -> go(type.member)
//      is Type.Map -> {
//        go(type.key)
//        go(type.value)
//      }
//      is Type.UntaggedUnion -> type.members.forEach(::go)
//      else -> {}
//    }
//  }
//
//  go(this)
//  return set.toList()
// }

fun Type.referencedShapeIds(): List<ShapeId> =
    when (this) {
      is Type.Ref -> listOf(shapeId)
      is Type.Set -> member.referencedShapeIds()
      is Type.List -> member.referencedShapeIds()
      is Type.Map -> key.referencedShapeIds() + value.referencedShapeIds()
      is Type.UntaggedUnion -> members.flatMap { it.referencedShapeIds() }
      else -> emptyList()
    }

sealed interface JsonRpcMethodType {
  object Request : JsonRpcMethodType

  object Notification : JsonRpcMethodType
}

data class Operation(
    val shapeId: ShapeId,
    val inputType: Type,
    val outputType: Type,
    val jsonRpcMethodType: JsonRpcMethodType,
    val jsonRpcMethod: String,
    val hints: List<Hint>
) {
  val name: String
    get() = shapeId.name
}

data class Field(val name: String, val type: Type, val required: Boolean, val hints: List<Hint>)

sealed interface EnumType<A> {
  object IntEnum : EnumType<Int>

  object StringEnum : EnumType<String>
}

data class EnumValue<A>(val name: String, val value: A, val hints: List<Hint>)

data class PolymorphicDataKind(val kind: String, val shape: ShapeId, val hints: List<Hint>)

sealed interface Hint {
  data class Documentation(val string: String) : Hint

  data class Deprecated(val message: String) : Hint

  data class JsonRename(val name: String) : Hint

  object Unstable : Hint
}
